### ISSUE TITLE
ImageRaster confusion with 3d textures

### ISSUE THREAD

**Issue by riccardobl** [jMonkeyEngine/jmonkeyengine#1021]
I want to split a 256x16 grid into a 16x16x16 3d texture.
This is my code
```java
Image img=new Image(Format.RGB8, 16, 16, 
	null,
	ColorSpace.Linear
);
img.addData(
	BufferUtils.createByteBuffer(16*16*16*
	(Format.RGB8.getBitsPerPixel()/8))
);
img.setDepth(16);
Texture3D tx=new Texture3D(img);
for(int i=0;i<16;i++){
	ImageRaster ir=ImageRaster.create(img,i);
	for(int x=0;x<16;x++){
		for(int y=0;y<16;y++){
                        [...]
			ir.setPixel(x,y,color);							
		}
	}
}

```

This is the exception i get when the ImageRaster is created on the second layer (it works fine for i=0)

```
java.lang.NullPointerException: null
        at com.jme3.texture.image.ByteOffsetImageCodec.writeComponents(ByteOffsetImageCodec.java:85) ~[bin/:?]
        at com.jme3.texture.image.DefaultImageRaster.setPixel(DefaultImageRaster.java:152) ~[bin/:?]
[...]
```

Following the exception 

--->

```java
        codec.writeComponents(getBuffer(), x, y, width, offset, components, temp);
```

--->

```java
 private ByteBuffer getBuffer(){
        if(buffer == null){
            this.buffer = image.getData(slice);
        }
        return buffer;
    }
```
--> image.getData(slice);
From the javadoc
> *  slice Which slice to use. Only applies to *3D images*, 2D image
     * arrays or cubemaps.

ImageRaster assumes that different layers of the 3d texture are stored in different ByteBuffers, while TextureUtil assumes that the layers are stored interleaved on a single bytebuffer.

```java
     if (target == GL2.GL_TEXTURE_3D) {
                gl2.glTexImage3D(target,
                                 level,
                                 format.internalFormat,
                                 width,
                                 height,
                                 depth,
                                 0,
                                 format.format,
                                 format.dataType,
                                 data);
    } 
```

**Comment by riccardobl** [comment]
Possible patch
*com.jme3.texture.image.DefaultImageRaster.java*
```patch
@@ -155,7 +155,14 @@
     
     private ByteBuffer getBuffer(){
         if(buffer == null){
-            this.buffer = image.getData(slice);
+		if(image.getDepth()>1){
+			int skip = image.getWidth() * image.getHeight() * codec.bpp * slice;
+			this.buffer = image.getData(0);
+			this.buffer.position(skip);
+			this.buffer = this.buffer.slice();
+		}else{
+            		this.buffer = image.getData(slice);
+		}
         }
         return buffer;
     }
```
I need some feedback on this since i don't know if it could break the engine elsewhere.

**Comment by riccardobl** [comment]
Fixed in d6950889f6db58b63cbdaae5e29cd2fb0a0d4452
